File: frontend/src/lib/utils/sanitize.ts
import DOMPurify from 'dompurify';

/**
 * Sanitize HTML from user-generated rich text (Tiptap review/comment content).
 * Allows safe formatting tags only — strips scripts, event handlers, and dangerous attributes.
 */
export function sanitizeHtml(dirty: string): string {
	return DOMPurify.sanitize(dirty, {
		ALLOWED_TAGS: [
			'p',
			'br',
			'strong',
			'b',
			'em',
			'i',
			'u',
			'ul',
			'ol',
			'li',
			'a',
			'span',
			'h2',
			'h3',
			'img',
			'table',
			'thead',
			'tbody',
			'tr',
			'th',
			'td',
		],
		ALLOWED_ATTR: ['href', 'target', 'rel', 'class', 'src', 'alt', 'width', 'height'],
		ALLOWED_URI_REGEXP: /^(?:(?:https?|mailto):|[^a-z]|[a-z+.-]+(?:[^a-z+.:-]|$))/i,
	});
}

File: backend/internal/core/services/sanitize.go
package services

import "github.com/microcosm-cc/bluemonday"

func newReviewPolicy() *bluemonday.Policy {
	p := bluemonday.NewPolicy()
	p.AllowElements("p", "br", "strong", "b", "em", "i", "u", "span", "ul", "ol", "li")
	p.AllowElements("h2", "h3")
	p.AllowElements("table", "thead", "tbody", "tr", "th", "td")

	p.AllowAttrs("href").OnElements("a")
	p.AllowAttrs("target").OnElements("a")
	p.AllowAttrs("rel").OnElements("a")
	p.RequireNoFollowOnLinks(false) // frontend already sets rel explicitly; don't fight it
	p.AllowStandardURLs()

	p.AllowAttrs("src", "alt", "width", "height").OnElements("img")
	p.AllowURLSchemes("http", "https")

	return p
}

var reviewPolicy = newReviewPolicy()

// sanitizeReview strips any HTML outside the perspective editor's supported
// tag set, mirroring frontend/src/lib/utils/sanitize.ts's DOMPurify config.
// This is defense-in-depth: the API must not trust HTML from non-browser
// clients that bypass the frontend's DOMPurify pass.
func sanitizeReview(html string) string {
	return reviewPolicy.Sanitize(html)
}
